### PR TITLE
Backports from https://github.com/paritytech/primitives/pull/15

### LINKED ISSUE #108: Add tests for `serde serialize/deserialize`

### PR DESCRIPTION
**Pull request by ordian**
Somehow these commits [1](https://github.com/paritytech/parity-common/commit/942fbf392b267c84520e4d8d890fc5a2c0cca8d9) [2](https://github.com/paritytech/parity-common/commit/13209866425e1705b2f38659b197d73b2ed4031d) [3](https://github.com/paritytech/parity-common/commit/90a3ada499a8d84009975e57e23ad03917d062ca) were lost in a migration. This PR brings them back.

Also fixes error message in deserialization (d4716c595640d2de52cd9ccc1fe1c2b9d77aa12b) and bumps versions.

Closes #108

### CODE REVIEW

File: ethbloom/Cargo.toml [hunk 1]
```diff
@@ -10,9 +10,9 @@ repository = "https://github.com/paritytech/parity-common"
 
 [dependencies]
 tiny-keccak = "1.4"
-crunchy = { version = "0.2.1", default-features = false, features = ["limit_256"] }
+crunchy = { version = "0.2", default-features = false, features = ["limit_256"] }
```
Comment (dvdplm) [inline ethbloom/Cargo.toml]: Not relevant to this PR, but I wonder what code we're unrolling to need 256 levels?

File: primitive-types/impls/serde/src/serialize.rs [hunk 2]
```diff
@@ -30,51 +54,38 @@ pub fn serialize_uint<S>(bytes: &[u8], serializer: S) -> Result<S::Ok, S::Error>
 		return serializer.serialize_str("0x0");
 	}
 
-	let hex: String = ::rustc_hex::ToHex::to_hex(bytes);
-	let has_leading_zero = !hex.is_empty() && &hex[0..1] == "0";
-	serializer.serialize_str(
-		&format!("0x{}", if has_leading_zero { &hex[1..] } else { &hex })
-	)
+	serializer.serialize_str(to_hex(slice, bytes, true))
```
Comment (dvdplm) [inline primitive-types/impls/serde/src/serialize.rs]: I don't want to re-review the changes here, but I do wonder why we're re-writing hex-ing code. Isn't `rustc_hex` doing a good job?

Comment (dvdplm) [reply primitive-types/impls/serde/src/serialize.rs]: I think the old code used `rustc_hex-1.0`; switching to `2.0` might have been enough to remove the heap allocations.

File: primitive-types/impls/serde/src/serialize.rs [hunk 3]
```diff
@@ -30,51 +54,38 @@ pub fn serialize_uint<S>(bytes: &[u8], serializer: S) -> Result<S::Ok, S::Error>
...
 
-impl fmt::Display for ExpectedLen {
+impl<'a> fmt::Display for ExpectedLen<'a> {
 	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
 		match *self {
-			ExpectedLen::Any => write!(fmt, "even length"),
-			ExpectedLen::Exact(v) => write!(fmt, "length of {}", v * 2),
-			ExpectedLen::Between(min, max) => write!(fmt, "length between ({}; {}]", min * 2, max * 2),
+			ExpectedLen::Exact(ref v) => write!(fmt, "length of {}", v.len() * 2),
+            ExpectedLen::Between(min, ref v) => write!(fmt, "length between ({}; {}]", min * 2, v.len() * 2),
```
Comment (dvdplm) [inline primitive-types/impls/serde/src/serialize.rs]: This whitespace looks odd.

File: ethbloom/Cargo.toml (hunk 1 above)
Comment (ordian) [reply ethbloom/Cargo.toml]: I guess that depends on a user-defined uint type size. That way we support numbers up to 256 * 8 bytes.

File: primitive-types/impls/serde/src/serialize.rs (hunk 2 above)
Comment (ordian) [reply primitive-types/impls/serde/src/serialize.rs]: Correct me if I'm wrong, but it seems like `rustc-hex 2.0` [uses](https://github.com/debris/rustc-hex/blob/3e773fee3992d8a2ecfcc22cc99141c2a6bf76c9/src/lib.rs#L182) `FromIterator<u8>` as a generic return type, but this trait is [only implemented](https://doc.rust-lang.org/std/iter/trait.FromIterator.html#implementors) for boxed types.

**Comment by ordian** [comment]
Published `impl-serde-v0.2.0`, `primitive-types-v0.2.4`, `ethereum-types-v0.5.3-beta.2`, `ethbloom-v0.6.3`, added git tags.

**Pull request merged**